ext/curl/share.c

```diff
@@ -134,6 +136,159 @@ PHP_FUNCTION(curl_share_strerror)
 }
 /* }}} */
 
+/**
+ * Initialize a persistent curl share handle with the given share options, reusing an existing one if found.
+ *
+ * Throws an exception if the share options are invalid.
+ */
+PHP_FUNCTION(curl_share_init_persistent)
+{
+	zval *share_opts = NULL, *entry = NULL;
+	zend_ulong persistent_id = 0;
+
+	php_curlsh *sh = NULL;
+
+	CURLSHcode error = 0;
+
+	ZEND_PARSE_PARAMETERS_START(1, 1)
+		Z_PARAM_ARRAY(share_opts)
+	ZEND_PARSE_PARAMETERS_END();
+
+	ZEND_HASH_FOREACH_VAL(Z_ARRVAL_P(share_opts), entry) {
+		ZVAL_DEREF(entry);
+
+		bool       failed = false;
+		zend_ulong option = zval_try_get_long(entry, &failed);
+
+		if (failed) {
+			zend_argument_type_error(1, "must contain only integer values, %s given", zend_zval_value_name(entry));
+			goto error;
+		}
+
+		switch (option) {
+			// Disallowed options
+			case CURL_LOCK_DATA_COOKIE:
+				zend_argument_value_error(1, "must not contain CURL_LOCK_DATA_COOKIE because it is unsafe");
+				goto error;
+
+			// Allowed options
+			case CURL_LOCK_DATA_DNS:
+				persistent_id |= 1 << 0;
+				break;
+			case CURL_LOCK_DATA_SSL_SESSION:
+				persistent_id |= 1 << 1;
+				break;
+			case CURL_LOCK_DATA_CONNECT:
+				persistent_id |= 1 << 2;
+				break;
+			case CURL_LOCK_DATA_PSL:
+				persistent_id |= 1 << 3;
+				break;
+
+			// Unknown options
+			default:
+				zend_argument_value_error(1, "must contain only CURL_LOCK_DATA_* constants");
+				goto error;
+		}
+	} ZEND_HASH_FOREACH_END();
+
+	// Construct a property field for the CurlSharePersistentHandle object with the enabled share options.
+	zval share_opts_prop;
+	array_init(&share_opts_prop);
+
+	if (persistent_id & 1 << 0) {
+		add_next_index_long(&share_opts_prop, CURL_LOCK_DATA_DNS);
+	}
+
+	if (persistent_id & 1 << 1) {
+		add_next_index_long(&share_opts_prop, CURL_LOCK_DATA_SSL_SESSION);
+	}
+
+	if (persistent_id & 1 << 2) {
+		add_next_index_long(&share_opts_prop, CURL_LOCK_DATA_CONNECT);
+	}
+
+	if (persistent_id & 1 << 3) {
+		add_next_index_long(&share_opts_prop, CURL_LOCK_DATA_PSL);
+	}
+
+	object_init_ex(return_value, curl_share_persistent_ce);
+
+	zend_update_property(
+	    curl_share_persistent_ce,
+		Z_OBJ_P(return_value),
+		"options",
+		sizeof("options") - 1,
+		&share_opts_prop
+	);
+
+	zval_ptr_dtor(&share_opts_prop);
+
+	sh = Z_CURL_SHARE_P(return_value);
+
+	if (persistent_id) {
```

Comment:
We probably should handle an empty options array early by throwing ("must not be empty") and then this check can be removed. I don't think providing an empty options array is ever useful, no?

Reply:
I agree, and throwing an exception would make it clearer to the user that it's not going to do anything otherwise.